### PR TITLE
blockchain: Infrastructure to manage block index.

### LINKED ISSUE #913: Improve stake difficulty calculation during startup
I'm tracking down startup time for dcrd. The biggest offender at the moment for my install seems to be this:

https://github.com/decred/dcrd/blob/master/blockchain/thresholdstate.go#L276

Which causes a traversal of all blocks in the chain.

I have a POC for swapping this (loading ancestors by iterating back) by a call to loading the node by height, but wanted to get see whether there was a particular reason to traverse the nodes backwards?

Only potential issue I can think of at the moment is if the chain pointing to the initial `prevNode` is a side chain, but I don't think a side chain can grow > `confirmationWIndow`.

@marcopeereboom (or someone else) could please give me some insight on this?

### PR DESCRIPTION
**Pull request by davecgh**
## NOTE: It is not possible to downgrade after running this PR. If you plan to test it, please make a copy of your data directory first.

~**This PR requires #1067.**~ (now merged)

This adds infrastructure necessary for `blockchain` to independently manage and persist a full block index in a bucket.  The purpose of this infrastructure is to significantly speed up startup time which currently has to load a bunch of the most recent blocks in order to rebuild the stake state and also to pave the way towards being able to have the full block index in memory allows for a lot of optimizations and greatly simplified code.

As a part of the new block index infrastructure it adds a new field to the block node struct named status that consists of bit flags for keeping tracking of the validation state of each block.  The status is only stored in memory as of this commit, but it will be stored as a part of the block index in the upcoming block index migration code.

Since the field will be updated after node creation, it also introduces some new functions for interacting with the status field in a concurrent-safe fashion.

The status code is largely based on upstream commits 2492af0 and fb0d13c; however, it does not include some of the logic in the reorganization paths those commits include as that approach doesn't match the intended direction this package is moving towards.

This also includes code to migrate the existing block index in `ffldb` to the new format managed by the `blockchain` package and updates the code to use the new infrastructure.

Finally, it includes a full suite of tests to ensure proper serialization and deserialization functionality.

---

### Testing Notes

As of this PR, the expected behavior is that there is a single migration that takes a while to complete, but subsequent startups should be significantly faster than the current code on master.  In particular, the point just after the chain state is reported where it is loading the initial portions of the block index.

As the warning above notes, if you try to run an older software version after this migration has completed, you will get an error message similar to `Unable to start server on [:19108]: the current blockchain database is no longer compatible with this version of the software (3 > 2).`

---

TODO:

- [x] ~Introduce a status to block nodes which tracks various state information such as whether or not the block has been stored to the database yet and whether or not it has already passed or failed validation~
- [x] ~Persist the status information to the new block index~

Closes #913.

### CODE REVIEW

**Comment by davecgh** [comment]
I've updated this to include the database migration code and to make use of the infrastructure.  It is split into more than one commit for ease of review.

**Comment by raedah** [comment]
Tested on testnet and mainnet. Rebuilt the dcrdata db with this from scratch, which pulls all data out of dcrd via rpc. No problems found. Dcrd startup time went from 20 second to 3 seconds.

**Comment by davecgh** [comment]
Rebased to latest master and resolved conflicts.

**Comment by davecgh** [comment]
I've updated this to include another commit which includes the new block validation status tracking and updated the PR description accordingly.

**Comment by davecgh** [comment]
The migration commit has now been updated to include persisting the new status information as a part of the block index record.  Note that if you previously tested this PR, the previously migrated database will have to be removed as it is not compatible with the latest changes.

**Comment by davecgh** [comment]
This is now ready for final review and testing.

File: blockchain/chain.go [hunk 1]
```diff
@@ -745,6 +745,12 @@ func (b *BlockChain) getReorganizeNodes(node *blockNode) (*list.List, *list.List
 		return detachNodes, attachNodes, nil
 	}
 
+	// Don't allow a reorganize to a descendant of an known invalid block.
```
Comment (dnldd) [inline blockchain/chain.go]: correction: an -> a

Comment (davecgh) [reply blockchain/chain.go]: Fixed

**Comment by dajohi** [comment]
tOK - mined on testnet2

**Comment by davecgh** [comment]
Some additional timing information:



## Time to migrate `testnet2` on a 7200 RPM HDD:
```
CHAN: Done upgrading block index.  Total entries: 241152 in 221 seconds 
```

Startup time before migration: ~2m34s
Startup time after migration:  ~9s


## Time to migrate `mainnet` on an SSD:
```
CHAN: Done upgrading block index.  Total entries: 215262 in 59 seconds
```

Startup time before migration: ~21s
Startup time after migration: ~2.7s

File: blockchain/accept.go [hunk 2]
```diff
@@ -126,13 +126,37 @@ func (b *BlockChain) maybeAcceptBlock(block *dcrutil.Block, flags BehaviorFlags)
 		return false, ruleError(ErrMissingParent, str)
 	}
 
+	// The is no need to validate the block if an ancestor is already known
```
Comment (chappjc) [inline blockchain/accept.go]: The --> There

File: blockchain/upgrade.go [hunk 3]
```diff
@@ -172,6 +197,201 @@ func upgradeToVersion2(db database.DB, chainParams *chaincfg.Params, dbInfo *dat
...
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	// doBatch contains the primary logic for upgrading the block index from
+	// version 1 to 2 in batches.  This is done because attempting to migrate in
+	// a single database transaction could result in massive memory usage and
+	// could potentially crash on  many systems due to ulimits.
```
Comment (chappjc) [inline blockchain/upgrade.go]: "on  many" --> "on many"

File: blockchain/upgrade.go [hunk 4]
```diff
@@ -172,6 +197,201 @@ func upgradeToVersion2(db database.DB, chainParams *chaincfg.Params, dbInfo *dat
...
+		if v2BlockIdxBucket == nil {
+			return 0, fmt.Errorf("bucket %s does not exist", v2BucketName)
+		}
+
+		hashIdxBucket := meta.Bucket(hashIdxBucketName)
+		if hashIdxBucket == nil {
+			return 0, fmt.Errorf("bucket %s does not exist", hashIdxBucketName)
+		}
+
+		// Migrate block index entries  so long as the max number of entries for
```
Comment (chappjc) [inline blockchain/upgrade.go]: "entries  so" --> "entries so"

File: blockchain/accept.go (hunk 2 above)
Comment (davecgh) [reply blockchain/accept.go]: Fixed

File: blockchain/upgrade.go (hunk 3 above)
Comment (davecgh) [reply blockchain/upgrade.go]: Fixed

File: blockchain/upgrade.go (hunk 4 above)
Comment (davecgh) [reply blockchain/upgrade.go]: Fixed

**Comment by raedah** [comment]
Tested this again with the same procedure as before, and found no problems.

**Pull request merged**